src/promise.h

```diff
@@ -1,28 +1,96 @@
 #pragma once
 
 #include <kj/async.h>
+#include <kj/debug.h>
 #include <rust/cxx.h>
 
+// These types are shared with rust
 namespace kj_rs {
 
 using OwnPromiseNode = kj::_::OwnPromiseNode;
 
-void own_promise_node_drop_in_place(OwnPromiseNode*);
-
-// https://github.com/dtolnay/cxx/blob/86cd652c06c5cb4c2e24d3ab555cf707b4ae0883/src/cxx.cc#L518
 namespace repr {
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wreturn-type-c-linkage"
+
+// ::cxx::private::PtrLen
 struct PtrLen final {
-  void *ptr;
-  std::size_t len;
+  void *ptr = nullptr;
+  std::size_t len = 0;
+};
+
+extern "C" {
+repr::PtrLen cxxbridge1$exception(const char *, std::size_t len) noexcept;
+}
+
+// ::cxx::private::Result
+struct Result final {
+  PtrLen err = {};
+  inline static Result ok() { return {}; }
+  inline static Result error(kj::Exception &e);
 };
+
+// ::kj::promise::UnwrapCallback
+using UnwrapCallback = Result (*)(const void /* kj::_::PromiseNode */ *node,
+                                  void /* T */ *ret);
+// ::kj::promise::KjPromiseNodeImpl
+struct KjPromiseNodeImpl {
+  template <typename T> inline KjPromiseNodeImpl(kj::Promise<T> &&p);
```

Comment:
Silly KJ style nit:
```suggestion
  template <typename T> inline KjPromiseNodeImpl(kj::Promise<T>&& p);
```

I see you also have `*` next to pointer identifiers everywhere, too ... maybe we should just wait until we can setup clang-format to run on this codebase.

Reply:
I didn't think about it at all: I just hit "format" in vscode and I have 0 idea what it does.

Here's clang-format anyway: https://github.com/capnproto/kj-rs/pull/19

Reply:
I'll reformat my PR after that is landed.

Reply:
rebased and reformatted